src/main.rs

```diff
@@ -116,28 +134,84 @@ fn run() -> Result<(), Report> {
-                .suggestion("Quickfixing a merge commit is not supported. If you meant to do this please file a ticket with your usecase.");
-        };
-        let head_1 = fix_commit.parent(0)?;
-        repo.reset(&head_1.as_object(), ResetType::Hard, None)?;
+    Ok(())
+}
+
+fn remove_commit(repo: &mut Repository, opts: &Opt, is_dirty: bool) -> Result<(), Report> {
+    if is_dirty {
+        // We should only get here if stashing is enabled
+        assert!(opts.stash);
+
+        // Stash everything
+        repo.stash_save(&repo.signature().unwrap(), "auto-stash: quickfix", None)?;
+    }
+    // Equivalent to git reset --hard HEAD~1
+    let head_1 = repo.head()?.peel_to_commit()?.parent(0)?;
+    repo.reset(&head_1.as_object(), ResetType::Hard, None)?;
+    drop(head_1);
+
+    if is_dirty {
+        // apply the staged changes
+        repo.stash_apply(0, None)?;
     }
 
+    Ok(())
+}
+
+fn push_new_commit(_repo: &Repository, opts: &Opt) -> Result<(), Report> {
     // TODO: Use git2 instead of Command.
-    if opts.push {
-        log::info!("Pushing new branch to origin.");
-        let status = Command::new("git")
-            .args(&["push", "--set-upstream", "origin", &opts.branch])
-            .status()?;
-        if !status.success() {
-            eyre!("Failed to run git push. {}", status);
-        } else {
-            log::info!("Git push succeeded");
+    log::info!("Pushing new branch to origin.");
+    let status = Command::new("git")
+        .args(&["push", "--set-upstream", "origin", &opts.branch])
+        .status()?;
+    if !status.success() {
+        eyre!("Failed to run git push. {}", status);
+    } else {
+        log::info!("Git push succeeded");
+    }
+
+    Ok(())
+}
+
+fn can_commit_be_kept(repo: &Repository, opts: &Opt) -> Result<bool, Report> {
+    if opts.keep {
+        // If we keep the commit, the repository state does not matter
+        Ok(true)
+    } else {
+        // Make sure that no rebase / cherry-pick / merge is in progress
+        let state = repo.state();
+        if state != RepositoryState::Clean {
+            return Err(eyre!(
+                "The repository is currently not in a clean state ({:?}).",
+                state
+            ));
         }
+
+        let is_dirty = !repo.statuses(None)?.is_empty();
+        if is_dirty && !opts.stash {
+            // Make sure that the work directory has no changes and nothing is staged
+            return Err(eyre!(
+                "The repository is dirty, aborting. Consider auto-stashing your changes with --stash."
+            ));
+        }
+        Ok(is_dirty)
+    }
+}
+
+fn run() -> Result<(), Report> {
+    let opts = Opt::from_args();
+    let mut repo = Repository::open_from_env()?;
+
+    let is_dirty = can_commit_be_kept(&repo, &opts)?;
+
+    cherrypick_commit_onto_new_branch(&repo, &opts)?;
+
+    if !opts.keep {
+        remove_commit(&mut repo, &opts, is_dirty)?
```

Comment:
Maybe I misunderstand this, but shouldn't it be `if !opts.keep && is_dirty` ? I see that it would potentially stash if `is_dirty` is true and `--stash` has been provided. That seems to be mixing responsibilities a bit.

Reply:
The logic now says that if the user does not specify `--keep`, the command should always try to remove the commit.
Removing the commit will raise an error if the working directory is dirty and `--stash` is not given.

I think the latest commit can be removed if either
1. the working directory is clean, aka `!is_dirty`, or
2. `is_dirty` and `opts.stash` is activated.

Changing the line to
```rust
if !opts.keep && is_dirty {
```
would cause the 1. case to fail, right?

Would you prefer to inline the `remove_commit` function here to make the "stashing" more visible?